File: tools/dice/src/dice/parser.py
"""Dice expression parser.

Grammar (Phase 1):
    expression := term ( ('+' | '-') term )*
    term       := dice | constant
    dice       := <count>d<sides>[k(h|l)<n>]
    constant   := <integer>
"""

from __future__ import annotations

import re
import secrets
from dataclasses import dataclass
from typing import Literal, Optional, Union


@dataclass(frozen=True)
class DiceTerm:
    count: int
    sides: int
    sign: int  # +1 or -1
    keep: Optional[tuple[Literal["h", "l"], int]]


@dataclass(frozen=True)
class ConstantTerm:
    value: int
    sign: int  # +1 or -1


Term = Union[DiceTerm, ConstantTerm]


# Captures a signed term: optional leading +/-, then a dice or constant body.
_TERM_RE = re.compile(
    r"""
    \s*
    (?P<sign>[+-])?\s*
    (?:
        (?P<count>\d+)d(?P<sides>\d+)
        (?:k(?P<keep>[hl])(?P<keep_n>\d+))?
        |
        (?P<const>\d+)
    )
    """,
    re.VERBOSE,
)


def parse_expression(expression: str) -> list[Term]:
    """Parse a dice expression into a list of signed Terms.

    Raises ValueError on empty or unparseable input.
    """
    if not expression or not expression.strip():
        raise ValueError("empty dice expression")

    pos = 0
    terms: list[Term] = []
    first = True
    s = expression.strip()

    while pos < len(s):
        m = _TERM_RE.match(s, pos)
        if m is None or m.end() == pos:
            raise ValueError(f"unparseable at position {pos}: {s[pos:]!r}")

        sign_str = m.group("sign")
        if first and sign_str is None:
            sign = 1
        elif sign_str is None:
            # No sign between terms is invalid (e.g., "1d6 1d8")
            raise ValueError(f"missing operator at position {pos}: {s[pos:]!r}")
        else:
            sign = 1 if sign_str == "+" else -1

        if m.group("const") is not None:
            terms.append(ConstantTerm(value=int(m.group("const")), sign=sign))
        else:
            count = int(m.group("count"))
            sides = int(m.group("sides"))
            keep = None
            if m.group("keep") is not None:
                keep = (m.group("keep"), int(m.group("keep_n")))
            terms.append(
                DiceTerm(count=count, sides=sides, sign=sign, keep=keep)
            )

        pos = m.end()
        first = False
        # Skip whitespace before the next operator
        while pos < len(s) and s[pos].isspace():
            pos += 1

    if not terms:
        raise ValueError(f"no terms parsed from {expression!r}")
    return terms


def _roll_die(sides: int) -> int:
    return secrets.randbelow(sides) + 1


def _apply_keep(
    rolls: list[int], keep: Optional[tuple[Literal["h", "l"], int]]
) -> tuple[list[int], list[int]]:
    """Return (kept, dropped) lists. If keep is None, all rolls kept."""
    if keep is None:
        return list(rolls), []
    direction, n = keep
    if n >= len(rolls):
        return list(rolls), []
    sorted_indexed = sorted(enumerate(rolls), key=lambda p: p[1])
    if direction == "h":
        kept_indices = {i for i, _ in sorted_indexed[-n:]}
    else:
        kept_indices = {i for i, _ in sorted_indexed[:n]}
    kept = [r for i, r in enumerate(rolls) if i in kept_indices]
    dropped = [r for i, r in enumerate(rolls) if i not in kept_indices]
    return kept, dropped


def _format_dice_term(term: DiceTerm) -> str:
    base = f"{term.count}d{term.sides}"
    if term.keep is not None:
        direction, n = term.keep
        base += f"k{direction}{n}"
    return base


def roll(expression: str) -> dict:
    """Roll a parsed dice expression. Returns a structured result."""
    terms = parse_expression(expression)
    out_terms: list[dict] = []
    total = 0

    for term in terms:
        if isinstance(term, ConstantTerm):
            value = term.sign * term.value
            total += value
            out_terms.append(
                {
                    "type": "constant",
                    "value": value,
                    "sign": term.sign,
                }
            )
        else:
            rolls = [_roll_die(term.sides) for _ in range(term.count)]
            kept, dropped = _apply_keep(rolls, term.keep)
            term_value = term.sign * sum(kept)
            total += term_value
            out_terms.append(
                {
                    "type": "dice",
                    "expr": _format_dice_term(term),
                    "rolls": rolls,
                    "kept": kept,
                    "dropped": dropped,
                    "value": term_value,
                    "sign": term.sign,
                }
            )

    return {
        "expression": expression,
        "terms": out_terms,
        "total": total,
    }

File: app/services/dice/parser.rb
module Dice
  module Parser
    DiceTerm     = Data.define(:count, :sides, :sign, :keep)
    ConstantTerm = Data.define(:value, :sign)

    MAX_COUNT = 100
    MAX_SIDES = 10_000

    TERM_RE = /
      \A
      \s*
      (?<sign>[+-])?\s*
      (?:
        (?<count>\d+)d(?<sides>\d+)
        (?:k(?<keep>[hl])(?<keep_n>\d+))?
        |
        (?<const>\d+)
      )
    /x

    module_function

    def parse(expression)
      raise Dice::ParseError, "empty dice expression" if expression.nil? || expression.strip.empty?

      s = expression.strip
      pos = 0
      terms = []
      first = true

      while pos < s.length
        remaining = s[pos..]
        match = TERM_RE.match(remaining)
        raise Dice::ParseError, "unparseable at position #{pos}: #{remaining.inspect}" if match.nil?

        sign_str = match[:sign]
        sign =
          if first && sign_str.nil?
            1
          elsif sign_str.nil?
            raise Dice::ParseError, "missing operator at position #{pos}: #{remaining.inspect}"
          else
            sign_str == "+" ? 1 : -1
          end

        if match[:const]
          terms << ConstantTerm.new(value: match[:const].to_i, sign: sign)
        else
          count = match[:count].to_i
          sides = match[:sides].to_i
          raise Dice::ParseError, "count must be between 1 and #{MAX_COUNT}, got #{count}" if count < 1 || count > MAX_COUNT
          raise Dice::ParseError, "sides must be between 1 and #{MAX_SIDES}, got #{sides}" if sides < 1 || sides > MAX_SIDES

          keep = nil
          if match[:keep]
            keep_n = match[:keep_n].to_i
            raise Dice::ParseError, "keep count must be >= 1, got #{keep_n}" if keep_n < 1
            keep = [ match[:keep].to_sym, keep_n ]
          end

          terms << DiceTerm.new(count: count, sides: sides, sign: sign, keep: keep)
        end

        pos += match.end(0)
        first = false
        pos += 1 while pos < s.length && s[pos] == " "
      end

      raise Dice::ParseError, "no terms parsed from #{expression.inspect}" if terms.empty?

      terms
    end
  end
end
